torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -367,85 +388,690 @@ bool PropagateTensorShapeOnNode(Node * node, bool insert_expands,
+  //   tensor inputs  : 1
+  //   tensor outputs : 1
+  // Additionally:
+  //   - First input should be the only tensor input
+  static const register_formula_for all_reduce_ops_with_integer_upcast {{
+    "aten::sum(Tensor self) -> Tensor",
+    "aten::prod(Tensor self) -> Tensor",
+  }, [](Node * node) -> type_vec_t {
+    if (auto type = node->input(0)->type()->cast<TensorType>()) {
+      return {at::isFloatingType(type->scalarType()) ? type : type->toScalarType(at::kLong)};
+    }
+    return {};
+  }};
+
+  static const auto multidim_reduce_with_postprocess =
+    [](Node * node, size_t num_reduced_dim, bool upcast_integer) -> type_vec_t {
+      auto maybe_keepdim = node->get<int64_t>(attr::keepdim);
+      if (!maybe_keepdim) return {};
+      if (auto type = node->input(0)->type()->cast<TensorType>()) {
+        if (upcast_integer && !at::isFloatingType(type->scalarType())) {
+          type = type->toScalarType(at::kLong);
         }
-        return nullptr;
+        if (*maybe_keepdim) {
+          return {type};
+        } else if (type->dim() > num_reduced_dim) {
+          return {type->withDim(type->dim() - num_reduced_dim)};
+        }
+      }
+      return {};
+    };
+
+  // Requirements:
+  //   dims           : preserved if keepdim == false, 1 smaller otherwise
+  //   scalar type    : preserved for first output, byte/uint8 for second output if exists
+  //   device         : preserved
+  //   tensor inputs  : 1
+  //   tensor outputs : 1 or 2
+  // Additionally:
+  //   - First input should be the only tensor input
+  //   - Has a bool keepdim argument
+  static const register_formula_for dim_reduce_ops {{
+    "aten::argmax(Tensor self, int dim, int keepdim) -> Tensor",
+    "aten::argmin(Tensor self, int dim, int keepdim) -> Tensor",
+    "aten::max_values(Tensor self, int dim, int keepdim) -> Tensor",
+    "aten::min_values(Tensor self, int dim, int keepdim) -> Tensor",
+    "aten::mean(Tensor self, int dim, int keepdim) -> Tensor",
+    "aten::norm(Tensor self, Scalar p, int dim, int keepdim) -> Tensor",
+    "aten::std(Tensor self, int dim, int unbiased, int keepdim) -> Tensor",
+    "aten::var(Tensor self, int dim, int unbiased, int keepdim) -> Tensor",
+    "aten::logsumexp(Tensor self, int dim, int keepdim) -> Tensor",
+    "aten::all(Tensor self, int dim, int keepdim) -> Tensor",
+    "aten::any(Tensor self, int dim, int keepdim) -> Tensor",
+
+    // Ops returning indices as second output
+    "aten::kthvalue(Tensor self, int k, int dim, int keepdim) -> (Tensor, Tensor)",
+    "aten::max(Tensor self, int dim, int keepdim) -> (Tensor, Tensor)",
+    "aten::min(Tensor self, int dim, int keepdim) -> (Tensor, Tensor)",
+    "aten::median(Tensor self, int dim, int keepdim) -> (Tensor, Tensor)",
+    "aten::mode(Tensor self, int dim, int keepdim) -> (Tensor, Tensor)",
+  }, [](Node * node) -> type_vec_t {
+    auto output_types = multidim_reduce_with_postprocess(node, /*num_reduce_dim=*/1, /*integer_upcast=*/false);
```

Comment:
comment: multidim_reduce_with_postprocess only returns a single element, even when the node has two outputs. 
This was confusing to me and I thought this code was wrong.